src/bin/pred.rs

```diff
@@ -323,6 +370,111 @@ pub fn sampler(rng: &mut Pcg64, num: u64, denom: u64, num_samples: usize) -> Vec
     ret
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum Z3<'a> {
+    Z3Bool(z3::ast::Bool<'a>),
+    Z3Real(z3::ast::Real<'a>),
+}
+
+impl<'a> Z3<'a> {
+    fn get_z3bool(&self) -> Option<z3::ast::Bool<'a>> {
+        match self {
+            Z3::Z3Bool(b) => Some(b.clone()),
+            Z3::Z3Real(_) => None,
+        }
+    }
+
+    fn get_z3rat(&self) -> Option<z3::ast::Real<'a>> {
+        match self {
+            Z3::Z3Bool(_) => None,
+            Z3::Z3Real(n) => Some(n.clone()),
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! pushbuf {
+    ($bu:ident, $ct:ident, $a:ident, $b:ident, $op:ident, $get:ident, $type:ident) => {
+        $bu.push(Z3::Z3Bool(z3::ast::$type::$op(
+            $ct,
+            &[
+                &$bu[usize::from(*$a)].$get().unwrap(),
+                &$bu[usize::from(*$b)].$get().unwrap(),
+            ],
+        )))
+    };
+    ($bu:ident, $a:ident, $b:ident, $op:ident, $get:ident, $type:ident) => {
+        $bu.push(Z3::Z3Bool(z3::ast::$type::$op(
+            &$bu[usize::from(*$a)].$get().unwrap(),
+            &$bu[usize::from(*$b)].$get().unwrap(),
+        )))
+    };
+    ($bu:ident, $a:ident, $op:ident, $get:ident, $type:ident) => {
+        $bu.push(Z3::Z3Bool(z3::ast::$type::$op(
+            &$bu[usize::from(*$a)].$get().unwrap(),
+        )))
+    };
+}
+
+fn egg_to_z3<'a>(ctx: &'a z3::Context, expr: &[Pred]) -> Z3<'a> {
+    let mut buf: Vec<Z3> = vec![];
+    for node in expr.as_ref().iter() {
+        match node {
+            Pred::Lit(lit) => match lit {
+                Constant::Bool(b) => buf.push(Z3::Z3Bool(z3::ast::Bool::from_bool(ctx, *b))),
+                Constant::Num(n) => buf.push(Z3::Z3Real(z3::ast::Real::from_real(
+                    ctx,
+                    (n.numer()).to_i32().unwrap(),
+                    (n.denom()).to_i32().unwrap(),
+                ))),
+            },
+            Pred::BVar(bv) => buf.push(Z3::Z3Bool(z3::ast::Bool::new_const(ctx, bv.to_string()))),
+            Pred::NVar(nv) => buf.push(Z3::Z3Real(z3::ast::Real::new_const(ctx, nv.to_string()))),
+            Pred::Lt([a, b]) => pushbuf!(buf, a, b, lt, get_z3rat, Real),
+            Pred::Leq([a, b]) => pushbuf!(buf, a, b, le, get_z3rat, Real),
+            Pred::Gt([a, b]) => pushbuf!(buf, a, b, gt, get_z3rat, Real),
+            Pred::Geq([a, b]) => pushbuf!(buf, a, b, ge, get_z3rat, Real),
+            Pred::And([a, b]) => pushbuf!(buf, ctx, a, b, and, get_z3bool, Bool),
+            Pred::Or([a, b]) => pushbuf!(buf, ctx, a, b, or, get_z3bool, Bool),
```

Comment:
Why do these need `ctx` and the rest don't?

Reply:
I am not really sure. It's just how the z3 rust api is built. Some of them don't pass the context around 🤷‍♀️